tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,224 @@
+    {out_name} = {ctype}();
+}}
+        """.split(
+            "\n"
+        )
+    elif isinstance(t, ListType):
+        in_name = arg_name + "_list_in"
+        out_name = arg_name + "_list_out"
+        elem_name = arg_name + "_elem"
+        code = [f"const c10::List<c10::IValue> {in_name} = {ival}.toList();"]
+        res_expr, res_code = argumenttype_ivalue_convert(t.elem, elem_name, arg_name)
+        connector = "\n\t"
+        # handle list type with size, e.g., bool[4]
+        if isinstance(t.elem, BaseType) and t.elem.name == BaseTy.bool and t.size:
+            code.extend(
+                f"""
+{ctype} {out_name} = as_array<{res_expr.type.cpp_type(strip_ref=True)}, {t.size}>({in_name});
+            """.split(
+                    "\n"
+                )
+            )
+        # we have to use c10::List for optional element. e.g., Tensor?[] -> c10::List<c10::optional<at::Tensor>>
+        elif isinstance(t.elem, OptionalType):
+            code.extend(
+                f"""
+{ctype} {out_name};
+for (c10::IValue {elem_name}: {in_name}) {{
+    {connector.join(res_code)}
+    {out_name}.push_back({res_expr.expr});
+}}
+            """.split(
+                    "\n"
+                )
+            )
+        else:
+            # use ArrayRef as default.
+            vec_name = arg_name + "_vec"
+            code.extend(
+                f"""
+std::vector<{res_expr.type.cpp_type(strip_ref=True)}> {vec_name};
+for (c10::IValue {elem_name}: {in_name}) {{
+    {connector.join(res_code)}
+    {vec_name}.push_back({res_expr.expr});
+}}
+{ctype} {out_name}({vec_name});
+            """.split(
+                    "\n"
+                )
+            )
+    else:
+        raise Exception(f"Cannot handle type {t}. ival: {ival}, arg_name: {arg_name}")
+    expr = Expr(expr=out_name, type=nctype)
+    return expr, code
+
+
+# Generate code to call C++ unboxed kernel with argument variable names
+def generate_unboxed_kernel_call(
+        f: NativeFunction, sig: CppSignature, arguments: List[Expr]
+) -> List[str]:
+    use_tensor_options = any(
+        isinstance(a.nctype.type, BaseCType) and a.nctype.type.type == tensorOptionsT
+        for a in sig.arguments()
+    )
+    # find dispatch native function namespace
+    if use_tensor_options:
+        namespace = "torch"
+    else:
+        namespace = "at"
+    # handle void return type
+    ret_type: CType = cpp.returns_type(f.func.returns)
+    if isinstance(ret_type, BaseCType) and ret_type.type == voidT:
+        ret_str = ""
+        push_str = ""
+    else:
+        ret_str = "auto result_ = "
+        push_str = """
+pack(stack, std::move(result_));
+        """
+    arg_connector = ",\n\t"
+    expr_list: List[Expr] = translate(arguments, sig.arguments(), method=(Variant.method in f.variants))
+    arg_list = f"\n\t{arg_connector.join(e.expr for e in expr_list)}\n"
+    if Variant.method in f.variants:
+        function_call = f"""
+{ret_str}self_base.{sig.name()}({arg_list});
```

Comment:
I think we can even remove this function completely, and move it all directly into `ComputeUnboxingFunctions.__call__()`. The more individual `f-strings` you have, and the more scattered they are across different files, the harder it is to see what the final output of the codegen is. Maybe more of this can be merged into a single f-string, that you can look at and get a better idea of what the whole output is.

Reply:
(the translate() call and the actual string that calls the C++ API are the only places where I think the `CppSignature` is used, so I think it would also make that clearer).